Ajax comments

wsf246 opened the pull request:
**Problem:** 
adds ajax to comments
**Solution:** 
uses ujs to create update edit and destroy comments. Closes #92 
**Complications:** 
I don't think the comments controller is validating the current user correctly
**Changelog:** 
-adds ujs for create, edit, update, and destroy
-removes options for non-comment creator
-updates comments count

wsf246 committed:
make comments usable by ojs

wsf246 committed:
pull comments out into partial

wsf246 committed:
destroy comment with ajax

wsf246 committed:
edit comments with form

wsf246 committed:
cancel edit of comment

wsf246 committed:
change comments count

wsf246 committed:
limit edit and delete to comment creator

briankung commented on app/controllers/comments_controller.rb:
```diff
@@ -11,14 +11,16 @@ def edit
+        @reference = Reference.find(@comment.root.commentable_id)
         format.html { redirect_to :back, notice: 'Comment was successfully created.' }
         format.json { render :show, status: :created, location: @comment }
+        format.js { render 'cud.js.erb' }
```
Please rename this partial.

wsf246 replied:
done

briankung commented on app/controllers/comments_controller.rb:
```diff
@@ -28,22 +30,27 @@ def create
   def update
     respond_to do |format|
       if @comment.update(comment_params)
+        @reference = Reference.find(@comment.root.commentable_id)
```
If you're using `commentable_id`, you can just do `@reference = @comment.root.commentable`

wsf246 replied:
done

briankung commented on app/views/comments/cud.js.erb:
```diff
@@ -0,0 +1,6 @@
+$("#comments-<%= @reference.id %>").html("<%= escape_javascript(render 'papers/comments', reference: @reference) %>");
```
Instead of a class variable, let's have the partial take it as a local variable. Makes it easier to reuse later - you don't need to set up the controller to have a `@reference` variable, you just feed a reference into the partial.

wsf246 replied:
done

briankung commented on app/views/comments/cud.js.erb:
```diff
@@ -0,0 +1,6 @@
+$("li#comments-tab-<%= @reference.id %>").html("<%= escape_javascript(render 'papers/comments_tab', reference: @reference) %>");
+$('.toggle-reply').on('click', function(e) {
+  e.preventDefault();
+  $(e.target).parents("ul").next('.comment-reply').toggleClass('hidden');
```
We should probably find a better way to do this

wsf246 replied:
I agree.  You mean specifically this function or rewriting this in this render.  I'm not 100% sure why me rendering papers/comment_tab again stops the reply button from working, maybe because the js we have in comments.js only starts after the DOM has loaded?

briankung commented on app/views/papers/_comments_tab.html.erb:
```diff
@@ -0,0 +1,5 @@
+  comments_count = reference.comments.map {|c| c.self_and_descendants.length}.inject(0,&:+)
+  display_count = comments_count.zero? ? "" : "(#{comments_count})"
+%>
+<%= link_to "Comments#{display_count}", '', "data-detail-id" => "comments-#{reference.id}", class: " paper-detail-tab" %>
```
👍  to breaking this into a partial

briankung commented on app/views/comments/edit.js.erb:
```diff
@@ -0,0 +1,11 @@
+    $(e.target).parents("#comment-content<%= @comment.id %>").html( "<%= escape_javascript(@comment.content) %>" );
+  });
+;
+
```
This file is extremely difficult to read. Is there anything you can do to increase the readability of what you're doing here?

wsf246 replied:
done

wsf246 committed:
rename cud.js.erb

wsf246 committed:
use simpler reference with commentable

wsf246 committed:
use local instead of class variable

wsf246 committed:
make edit.js.erb easier to read

wsf246 commented:
I think that should be all the points.  Let me know what you think.

briankung commented on app/views/comments/reference_comments.js.erb:
```diff
@@ -0,0 +1,6 @@
+$("li#comments-tab-<%= reference.id %>").html("<%= escape_javascript(render 'papers/comments_tab', reference: reference) %>");
+$('.toggle-reply').on('click', function(e) {
+  e.preventDefault();
+  $(e.target).parents("ul").next('.comment-reply').toggleClass('hidden');
```
We should probably just give each comment reply its own id and select based on that somehow. We can do it later, though.

briankung commented on app/views/comments/edit.js.erb:
```diff
@@ -0,0 +1,13 @@
+  commentContentP.html("<%= escape_javascript(@comment.content) %>");
+});
+
+
```
This file is waayyyy more readable \o/

briankung reviewed (approved):
This looks good to me. You can make that minor change I suggested or not, up to you 👍

briankung commented:
Just be aware of the merge conflicts.

wsf246 committed:
target specific reply form

wsf246 committed:
move reference comment partials from papers to reference folder

wsf246 committed:
create comments using jquery for just the comments changed

wsf246 committed:
remove extra space

wsf246 commented on app/controllers/comments_controller.rb:
```diff
@@ -11,14 +11,16 @@ def edit
+        reference = @comment.root.commentable
         format.html { redirect_to :back, notice: 'Comment was successfully created.' }
         format.json { render :show, status: :created, location: @comment }
+        format.js { render( @comment.parent.nil? ? 'comment_on_reference.js.erb' : 'reply_to_comment.js.erb', locals: {reference: reference} )}
```
so this is the part I was struggling with yesterday, but I've got a solution and wanted your opinion on it.  Basically because commenting on a reference is different enough than replying to a comment, I seperated out the logic into 2 different jquery pieces.  There's some duplication, so I was considering joining these two things with an if statement in the jquery, but this seems cleaner and I feel like it makes more sense to make the controller do that.  Thoughts?

briankung replied:
I feel like it should only require one jquery partial.

Also, you and your ternary operators. When these ternary operator lines get really long, I recommend breaking them out into `if/else` statements.

For this one in particular, you can do 

``` ruby
format.js do
  if @comment.parent
    render 'reply_to_comment.js.erb', locals: {reference: reference}
  else
    render 'comment_on_reference.js.erb', locals: {reference: reference}
  end
end
```

Way more readable.

That said, I strongly feel like it shouldn't require two separate partials.

briankung replied:
Let's pair program on this one soon.

wsf246 commented on app/views/comments/reply_to_comment.js.erb:
```diff
@@ -0,0 +1,14 @@
+commentParentDiv = $("#comment-<%= @comment.parent_id %>");
+replyDiv = commentParentDiv.children('.reply');
```
the main difference between replying to a comment or a reference is the reply div wrapper and seeing if it exists already.

briankung replied:
I feel like this should not be the case...let's see what we can do instead.

wsf246 committed:
destroy comment

wsf246 committed:
remove reply div if only sibling

wsf246 committed:
refactor destroy

wsf246 committed:
update button

briankung commented on app/assets/javascripts/comments.js:
```diff
@@ -1,6 +1,7 @@
     e.preventDefault();
-    $(e.target).parents("ul").next('.comment-reply').toggleClass('hidden');
+    commentID = $(e.target).data('parent-comment-id');
+    $(e.target).parents("ul").next("#reply-form-"+commentID).toggleClass('hidden');
```
Since we're using an this id `#reply-form-"+commentID` we don't have to do the `$().parents().next()` thing. Since there should be only one - hence an ID - you can just grab the ID: `$("#reply-form-"+commentID)`

wsf246 replied:
I do this if I can because it limits the amount of jquerying so it doesn't have to hunt through the whole DOM to find it.  It's minor and not a big deal so I'm fine with whichever way.

briankung replied:
`$().parents().next()` is much more fragile if the HTML changes, which it probably will. I also think hunting through the DOM for an ID is pretty fast, considering there should be only one of them.

briankung commented on app/controllers/comments_controller.rb:
```diff
@@ -28,22 +30,27 @@ def create
   # DELETE /comments/1
   # DELETE /comments/1.json
   def destroy
+    reference = @comment.root.commentable
```
I forgot to note earlier that this will stop working if we add another model as commentable

briankung replied:
`¯\_(ツ)_/¯`

wsf246 replied:
lol, we'll cross that bridge when we get there.

briankung commented on app/views/comments/reply_to_comment.js.erb:
```diff
@@ -0,0 +1,14 @@
+replyDiv = commentParentDiv.children('.reply');
+
+if (replyDiv.length) {
+  replyDiv.children().last().after("<%= escape_javascript(render('comments/comment', comment: @comment)) %>");
```
I think you can just do `replyDiv.append()`

briankung commented on app/views/comments/reply_to_comment.js.erb:
```diff
@@ -0,0 +1,14 @@
+    "<%= escape_javascript(
+      content_tag( :div, render('comments/comment', comment: @comment), class: 'reply')
+    )%>"
+  );
```
Same here with `append()`

briankung commented on app/views/comments/destroy.js.erb:
```diff
@@ -0,0 +1,5 @@
+if (commentDiv.siblings().length == 0) { commentDiv.parent(".reply").remove() };
+commentDiv.remove();
+
+$("li#comments-tab-<%= reference.id %>").html("<%= escape_javascript(render 'references/comments_tab', reference: reference) %>");
```
Would you mind just decrement the counter instead of re-rendering the entire comments tab? You'd probably want to use JavaScript `parseInt`

briankung replied:
Then subtract one and reinsert it

briankung commented on app/views/comments/edit.js.erb:
```diff
@@ -0,0 +1,11 @@
+cancelLink.on('click', function(e) {
+  e.preventDefault();
+  commentContentP.html("<%= escape_javascript(@comment.content) %>");
+});
```
Is re-registering the handler necessary, or can you use the `$(document).on('event', 'selector' function)` live handler syntax?

wsf246 replied:
mmm, not sure, let's pair program on this one.

briankung commented:
Let's pair program on this one soon

wsf246 commented:
yea, let's and finish this guy up.  When is good for you?

wsf246 committed:
merge master and fix comment conflict

wsf246 committed:
update comment html structure

wsf246 committed:
update destroy comment using only comments count

wsf246 committed:
simplify update comment

wsf246 committed:
put id on comments count span

briankung committed:
Minor change

CommentID would actually camelcase to CommentId. This is true of Rails conventions, too. You may see places where you create a model that's SoAndSoID and it will convert to a snake-case of so_and_so_i_id

briankung commented on app/controllers/comments_controller.rb:
```diff
@@ -11,11 +11,33 @@ def edit
+          new_comment = @comment.parent.children.build
+          display_comment = false
+        else
+          parent_id = "r#{reference.id}"
```
This doesn't seem like a good idea. I'd prefer to switch based on a class or some way of indicating the type of element you want to anchor the comment to.

briankung commented on app/controllers/comments_controller.rb:
```diff
@@ -11,11 +11,33 @@ def edit
+              reference: reference,
+              parent_id: parent_id,
+              new_comment: new_comment,
+              display_comment: display_comment
```
I could go either way on this issue, but this seems like a lot of locals to be cramming into one partial.

wsf246 committed:
add class to child comment divs

wsf246 committed:
Merge branch 'ajaxComments' of github.com:briankung/projectcredo into ajaxComments

wsf246 commented on app/controllers/comments_controller.rb:
```diff
@@ -34,7 +36,8 @@ def create
               new_comment: new_comment,
-              display_comment: display_comment
+              display_comment: display_comment,
+              comments_class: comments_class
```
I added this class to each of the child comment divs to specify if it was for a reference or a comment.  Let me know if you can think of a more elegant way to do this, maybe putting it in the ID name itself.

briankung replied:
Let's revert this change, actually. Doesn't look like it serves a purpose.

wsf246 commented:
it's a lot of local variables to push to that partial, but they're largely due to the way closure tree is setup and how we need to build a comment.  I think it's fine for now and we can refactor or simplify later on, especially if we eventually move to disqus or something, probably more meaningful to push this working set of code for now.

briankung commented on app/views/comments/destroy.js.erb:
```diff
@@ -0,0 +1,7 @@
+comments = $("#comment-<%= @comment.id %>").find(".comment").length + 1;
+console.log(comments);
```
Debug statement

wsf246 replied:
shit, good catch.

wsf246 committed:
Remove debug line

wsf246 commented:
If this looks good, I'll merge.  I can leave some comments in the controller so we can come back to all those local variables getting pushed in later.

briankung reviewed (changes requested):
Let's use comment generations and sort order to target a comment or a generation to append a comment.

Basically it works like this:

```
comment A (generation 0)
    comment B (generation 1)
    comment C (generation 1)
        comment D (generation 2)
    comment E (generation 1)
        comment F (generation 2)
```

I'm writing up how to do this at the moment

briankung commented:
I'm just going to refactor this myself, if you don't mind

briankung committed:
Revert "add class to child comment divs"

This reverts commit f4ddca3c36a875be97e64187ebfd4ed0e915c54f.

briankung committed:
Whitespaces

briankung committed:
Moves the js specific stuff into the js block

briankung committed:
Breaks content tag out for readability

wsf246 commented:
yea go for it.

On Tue, Nov 1, 2016 at 1:36 PM, Brian Kung notifications@github.com wrote:

> I'm just going to refactor this myself, if you don't mind
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> https://github.com/briankung/projectcredo/pull/128#issuecomment-257654694,
> or mute the thread
> https://github.com/notifications/unsubscribe-auth/ACKsZPLiC16JHbrfjL9A2A_ED0c1Gw_Hks5q54bEgaJpZM4Kc3J7
> .

briankung committed:
Renames variable for clarity

briankung committed:
Inverts display comment field to make more sense
With a sensible default

wsf246 commented on app/views/comments/_comment.html.erb:
```diff
@@ -13,7 +13,7 @@
     <% end %>
   </ul>
-  <%= (render 'comments/form', comment: comment.children.build, display_comment_field: false) if current_user %>
+  <%= render('comments/form', comment: comment.children.build, hidden: true) if current_user %>
```
smart.

wsf246 commented on app/views/comments/_form.html.erb:
```diff
@@ -1,10 +1,12 @@
   <%= f.hidden_field :parent_id, value: comment.parent_id %>
-  <%= f.hidden_field :commentable_type, value: comment.commentable_type if display_comment_field %>
-  <%= f.hidden_field :commentable_id, value: comment.commentable_id if display_comment_field %>
+  <%= f.hidden_field :commentable_type, value: comment.commentable_type %>
```
good refactor

wsf246 commented on app/controllers/comments_controller.rb:
```diff
@@ -11,11 +11,30 @@ def edit
+            parent_id = @comment.parent_id
+            new_comment = @comment.parent.children.build
+          else
+            parent_id = "r#{reference.id}"
```
looks like you reverted back to having the "r" in the class.  Are you changing up this structure to match the generations?

wsf246 commented:
hey, just checking if you're still working on this or if you were looking to merge.  Looks like the list breaks on list show on the heroku pull request.

I was thinking if you were going to go for the generation class or id for the child comments div, you should probably include the reference ID.  It might be something like child-comments-ref-<%= ref.id %>-comment-gen-<%= gen %>.  The generation may not be granular enough either because gen 2 can sit under different comments depending on how you implement and if you are tracking based on the comment ID as well.

briankung commented:
I'm still working on this. Reverted to r to get things working for other
refactorings. About to get rid of it

On Thu, Nov 3, 2016, 12:10 PM wsf246 notifications@github.com wrote:

> hey, just checking if you're still working on this or if you were looking
> to merge. Looks like the list breaks on list show on the heroku pull
> request.
> 
> I was thinking if you were going to go for the generation class or id for
> the child comments div, you should probably include the reference ID. It
> might be something like child-comments-ref-<%= ref.id %>-comment-gen-<%=
> gen %>. The generation may not be granular enough either because gen 2 can
> sit under different comments depending on how you implement and if you are
> tracking based on the comment ID as well.
> 
> —
> You are receiving this because you commented.
> 
> Reply to this email directly, view it on GitHub
> https://github.com/briankung/projectcredo/pull/128#issuecomment-258189857,
> or mute the thread
> https://github.com/notifications/unsubscribe-auth/ACtGx3MzVZopcl3G-3VWuB1luGA2anjeks5q6gdFgaJpZM4Kc3J7
> .

briankung committed:
Redefines HTML for unified access to comments div

briankung committed:
Refactors / fixes edit js

briankung committed:
Refactors/fixes update.js

briankung committed:
Simplifies create.js for comment replies

briankung committed:
Removes reply-form id

briankung committed:
Removes unnecessary variable declaration

briankung committed:
Removes failure mode render

briankung committed:
Splits partial depending on whether it's top level or not

briankung commented:
So you were right way back when, they were different enough to split into separate partials. Sorry about that - you were correct.

Theoretically right now, we have a way to attach comments to a comments div with `comments[data-commentable-id=123]`, and use css selectors to pick elements with or without a `reference` class depending on whether it is or isn't a reference. If we could guarantee that ids were unique, then we wouldn't need to clarify whether or not it had a reference class. But that's kind of how I would want to select a comments container to add comments to. The `comments` class says it's a container for comments, and the combination of `data-commentable-id` and `reference` allow for picking a unique comments container, even though a comment and a reference might have the same ID.

briankung committed:
Clears textarea value upon submission

wsf246 commented on app/controllers/comments_controller.rb:
```diff
@@ -11,11 +11,20 @@ def edit
         format.json { render :show, status: :created, location: @comment }
+        format.js do
+          reference = @comment.root.commentable
+          is_top_level = !!@comment.commentable
```
!!!

briankung replied:
This is a trick to cast something into a boolean

wsf246 commented on app/controllers/comments_controller.rb:
```diff
@@ -40,10 +50,12 @@ def update
   # DELETE /comments/1
   # DELETE /comments/1.json
   def destroy
+    reference = @comment.root.commentable
```
might be worth doing a before_action for `reference = @comment.root.commentable`

wsf246 replied:
nevermind, it's not worth doing.

wsf246 commented on app/helpers/comments_helper.rb:
```diff
@@ -2,7 +2,12 @@ module CommentsHelper
+          :div,
+          comments_tree_for(nested_comments),
+          class: "comments",
+          'data-commentable-id' => comment.id
```
just out of understanding best practice, do you choose to create a custom data-id field instead of using the id field just to be specific?

briankung commented:
The ID field in HTML should be unique. Since I can't guarantee it's unique,
I am using a more explanatory field/field name

On Thu, Nov 3, 2016, 6:49 PM wsf246 notifications@github.com wrote:

> ## _@wsf246_ commented on this pull request.
> 
> In app/helpers/comments_helper.rb
> https://github.com/briankung/projectcredo/pull/128#pullrequestreview-7119547
> :
> 
> > @@ -2,7 +2,12 @@ module CommentsHelper
> >    def comments_tree_for(comments)
> >      comments.map do |comment, nested_comments|
> >        render 'comments/comment', comment: comment do
> > -        content_tag(:div, comments_tree_for(nested_comments), class: 'reply') if nested_comments.size > 0
> > -        content_tag(
> > -          :div,
> > -          comments_tree_for(nested_comments),
> > -          class: "comments",
> > -          'data-commentable-id' => comment.id
> 
> just out of understanding best practice, do you choose to create a custom
> data-id field instead of using the id field just to be specific?
> 
> —
> You are receiving this because you commented.
> 
> Reply to this email directly, view it on GitHub
> https://github.com/briankung/projectcredo/pull/128#pullrequestreview-7119547,
> or mute the thread
> https://github.com/notifications/unsubscribe-auth/ACtGx7-wqgsTBofD1XCxWHNH4MvUO-8fks5q6mTFgaJpZM4Kc3J7
> .

briankung commented:
Sure

On Thu, Nov 3, 2016, 6:47 PM wsf246 notifications@github.com wrote:

> ## _@wsf246_ commented on this pull request.
> 
> In app/controllers/comments_controller.rb
> https://github.com/briankung/projectcredo/pull/128#pullrequestreview-7119387
> :
> 
> > @@ -40,10 +50,12 @@ def update
> >    # DELETE /comments/1
> >    # DELETE /comments/1.json
> >    def destroy
> > -    reference = @comment.root.commentable
> 
> might be worth doing a before_action for reference =
> @comment.root.commentable
> 
> —
> You are receiving this because you commented.
> 
> Reply to this email directly, view it on GitHub
> https://github.com/briankung/projectcredo/pull/128#pullrequestreview-7119387,
> or mute the thread
> https://github.com/notifications/unsubscribe-auth/ACtGx6cQVoqC-rhMFD2evqgOS2tqq7Lqks5q6mRrgaJpZM4Kc3J7
> .

wsf246 commented on app/views/comments/create.js.erb:
```diff
@@ -0,0 +1,14 @@
+<%
+  comment_and_comments_container = render('comments/comment', comment: @comment) do
+    render('form', comment: @comment.children.build)
+    content_tag(:div, '', class: 'comments', 'data-commentable-id' => @comment.id)
```
fancy

briankung commented:
"Sure" was supposed to be a response for the before action suggestion

On Thu, Nov 3, 2016, 6:52 PM wsf246 notifications@github.com wrote:

> ## _@wsf246_ commented on this pull request.
> 
> In app/views/comments/create.js.erb
> https://github.com/briankung/projectcredo/pull/128#pullrequestreview-7119991
> :
> 
> > @@ -0,0 +1,14 @@
> > +<%
> > -  comment_and_comments_container = render('comments/comment', comment: @comment) do
> > -    render('form', comment: @comment.children.build)
> > -    content_tag(:div, '', class: 'comments', 'data-commentable-id' => @comment.id)
> 
> fancy
> 
> —
> You are receiving this because you commented.
> 
> Reply to this email directly, view it on GitHub
> https://github.com/briankung/projectcredo/pull/128#pullrequestreview-7119991,
> or mute the thread
> https://github.com/notifications/unsubscribe-auth/ACtGxzB7552j3R3fo7L6gOkMuZ0YRTFgks5q6mWZgaJpZM4Kc3J7
> .

wsf246 commented:
Gotcha, going back to what you were saying before.  So best practice, go
with just the ID field, but if you can't guarantee uniqueness, create a
data ID field explicitly.

On Thu, Nov 3, 2016 at 5:51 PM, Brian Kung notifications@github.com wrote:

> The ID field in HTML should be unique. Since I can't guarantee it's unique,
> I am using a more explanatory field/field name
> 
> On Thu, Nov 3, 2016, 6:49 PM wsf246 notifications@github.com wrote:
> 
> > ## _@wsf246_ commented on this pull request.
> > 
> > In app/helpers/comments_helper.rb
> > <https://github.com/briankung/projectcredo/pull/128#
> > pullrequestreview-7119547>
> > :
> > 
> > > @@ -2,7 +2,12 @@ module CommentsHelper
> > > def comments_tree_for(comments)
> > > comments.map do |comment, nested_comments|
> > > render 'comments/comment', comment: comment do
> > > - content_tag(:div, comments_tree_for(nested_comments), class: 'reply')
> > >   if nested_comments.size > 0
> > > - content_tag(
> > > - :div,
> > > - comments_tree_for(nested_comments),
> > > - class: "comments",
> > > - 'data-commentable-id' => comment.id
> > 
> > just out of understanding best practice, do you choose to create a custom
> > data-id field instead of using the id field just to be specific?
> > 
> > —
> > You are receiving this because you commented.
> > 
> > Reply to this email directly, view it on GitHub
> > <https://github.com/briankung/projectcredo/pull/128#
> > pullrequestreview-7119547>,
> > or mute the thread
> > <https://github.com/notifications/unsubscribe-auth/ACtGx7-
> > wqgsTBofD1XCxWHNH4MvUO-8fks5q6mTFgaJpZM4Kc3J7>
> > .
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> https://github.com/briankung/projectcredo/pull/128#issuecomment-258297727,
> or mute the thread
> https://github.com/notifications/unsubscribe-auth/ACKsZA0Wef8-tmYg8nE6f3Np-pD1wCF6ks5q6mVogaJpZM4Kc3J7
> .

wsf246 commented on app/views/comments/edit.js.erb:
```diff
@@ -0,0 +1,16 @@
+cancelLink = commentSubmitButton.next().find(".cancel-link");
+cancelLink.on('click', function(e) {
+  e.preventDefault();
+  commentContent.html(undoContent);
```
sweet
